k9mail/src/main/java/com/fsck/k9/controller/ImapSyncInteractor.java

```diff
@@ -0,0 +1,268 @@
+
+    private final Account account;
+    private final String folderName;
+    private LocalFolder localFolder;
+    private ImapFolder imapFolder;
+    private final  MessagingListener listener;
+    private final  MessagingController controller;
+
+    ImapSyncInteractor(Account account, String folderName, MessagingListener listener, MessagingController controller) {
+        this.account = account;
+        this.folderName = folderName;
+        this.listener = listener;
+        this.controller = controller;
+    }
+
+    void performSync(MessageDownloader messageDownloader) {
+        Exception commandException = null;
+
+        try {
+            Timber.d("SYNC: About to process pending commands for account %s", account.getDescription());
+
+            try {
+                controller.processPendingCommandsSynchronous(account);
+            } catch (Exception e) {
+                controller.addErrorMessage(account, null, e);
+                Timber.e(e, "Failure processing command, but allow message sync attempt");
+                commandException = e;
+            }
+
+            Timber.v("SYNC: About to get local folder %s and open it", folderName);
+            localFolder = SyncUtils.getOpenedLocalFolder(account, folderName);
+            localFolder.updateLastUid();
+
+            Store remoteStore = account.getRemoteStore();
+            Timber.v("SYNC: About to get remote folder %s", folderName);
+            Folder remoteFolder = remoteStore.getFolder(folderName);
+            if (!(remoteFolder instanceof ImapFolder)) {
+                throw new MessagingException("A non-IMAP account was provided to ImapSyncInteractor");
+            }
+            imapFolder = (ImapFolder) remoteFolder;
+
+            if (!SyncUtils.verifyOrCreateRemoteSpecialFolder(account, folderName, imapFolder, listener, controller)) {
+                return;
+            }
+
+            QresyncResponse qresyncResponse;
+            Timber.v("SYNC: About to open remote IMAP folder %s", folderName);
+            List<String> uids = localFolder.getAllMessageUids();
+            if (uids.size() == 0) {
+                qresyncResponse = imapFolder.open(Folder.OPEN_MODE_RW, localFolder.getUidValidity(),
+                        localFolder.getHighestModSeq());
+            } else {
+                long smallestUid = Long.parseLong(uids.get(uids.size() - 1));
+                qresyncResponse = imapFolder.open(Folder.OPEN_MODE_RW, localFolder.getUidValidity(),
+                        localFolder.getHighestModSeq(), smallestUid);
+            }
+
+            boolean qresyncEnabled = qresyncResponse != null;
+            List<String> expungedUids = new ArrayList<>();
+            if (Expunge.EXPUNGE_ON_POLL == account.getExpungePolicy()) {
+                Timber.d("SYNC: Expunging folder %s:%s", account.getDescription(), folderName);
+                if (qresyncEnabled) {
+                    expungedUids = imapFolder.expungeUsingQresync();
```

Comment:
so this code path is triggered when QresyncResponse is non-null, but doesn't actually depend on that response? I don't fully know the context here but that seems weird